packages/client/src/index.ts

```diff
@@ -11,10 +21,113 @@ export interface Service {
   url: URL;
 }
 
-// eslint-disable-next-line @typescript-eslint/no-unused-vars
-export function createClient(_serverURL: string): Client {
-  async function createSimulation() {
-    return { services: {} };
+interface Runnable<T> {
+  run(scope: Task): T;
+}
+
+interface Result<T> {
+  data: T;
+  errors?: GraphQLError[];
+}
+
+export function createClient(serverURL: string, webSocketImpl: unknown = WebSocket): Client {
```

Comment:
Could you not let typescript infer the type

```ts
export function createClient(serverURL: string, webSocketImpl = WebSocket): Client {
```

Reply:
We can, but don't we have a linting rule that exported functions should explicitly declare return type?

Reply:
Oh, you mean the webSocketImpl. The problem is that you can pass anything in there as `graphql-ws` treats it as `unknown`. We could define our own interface perhaps called `WebSocketImpl` but not sure how to go about typing it.

Reply:
totally up to  you if you want to change this, I just feel a bit uneasy about no typing for the `webSocketImpl`

You could do something like this also

```ts
export function createClient<W extends WebSocket = WebSocket>(serverURL: string, webSocketImpl: W): Client {
```

Reply:
Totally up to you if you want to change this.  If there are no other implementations yet, I would always leave just go with `WebSocket` and cross the other bridge when the need arises.

I've approved so this only minor and not necessary.

Reply:
The problem is that `window.WebSocket` and `ws.WebSocket` do not share a common type, so it does a bunch of weird duck typing https://github.com/enisdenjo/graphql-ws/blob/master/src/client.ts#L273-L291

I'd love to type it too, but not sure the best way.

Reply:
Something like this will work:

```ts
import ws from 'ws';

declare module 'ws' {
  interface WebSocket {
    CLOSED: number;
    CONNECTING: number;
    OPEN: number;
  }
}

export function createClient(serverURL: string, webSocketImpl: ws.WebSocket | WebSocket = WebSocket): Client {
```

It is using declaration merging to add the bits that tsc is complaining about.

We might put the `declare module 'ws'` bit in a `global.d.ts` file or `index.d.ts` so it gets picked  up automatically.

I'm off for a couple of days but I have approved and we could go over it on Friday if I am not understanding the suituation.

Reply:
What about this?

```ts
export interface WebSocketImpl {
  new(url: string): {
    send(data: string | ArrayBufferLike | Blob | ArrayBufferView): void;
  }
}
```

It matches both `ws.WebSocket` and `window.WebSocket`. basically A constructor function that returns an object with a `send()` method.

Reply:
looks good to me!  More than happy with either.

A public signature with an `unknown` field is a `known` problem :).